src/node/state.go

```diff
@@ -19,7 +15,29 @@ const (
 	Stop
 )
 
-func (s NodeState) String() string {
+type state int
+
+type nodeState2 struct {
+	cond *sync.Cond
+	lock sync.RWMutex
+	wip  int
+
+	state        state
+	getStateChan chan state
+	setStateChan chan state
+}
+
+func newNodeState2() *nodeState2 {
+	ns := &nodeState2{
+		cond:         sync.NewCond(&sync.Mutex{}),
+		getStateChan: make(chan state),
+		setStateChan: make(chan state),
+	}
+	go ns.mtx()
```

Comment:
This starts an additional goroutine which need to be scheduled, i.e. this approach is one of the most ineffective in respect of computational resources, in addition it doesn't solve the problem, which lies in fact, that waitRoutines() function must, in context of its current usage, not only wait until all started goFunc() terminate, but also prevent any new goFunc() started.

Reply:
I like memory synchronization through channels more.
Goroutine are very lightweight.

`waitRoutines` function should not block the execution of other goFunc()  functions. This is a bad design.
The task of `waitRoutines` function is to wait for the completion of all functions.

Reply:
very lightweight != zero weight ; 
The semantic of waitRoutines() usage is seems ensuring the node is actually switched into a new state, if so, it's better to solve that problem other way, more elegantly and with less cost.